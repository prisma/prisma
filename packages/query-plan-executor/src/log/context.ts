import { context } from '@opentelemetry/api'

import { Logger } from './logger.ts'

const LOG_CONTEXT = Symbol('LogContext')

/**
 * Gets the logger from the current OpenTelemetry context.
 *
 * @throws {Error} if no logger is found in the context or if the value is not a {@link Logger}
 */
export function getActiveLogger(): Logger {
  const logger = context.active().getValue(LOG_CONTEXT)

  if (!logger) {
    throw new Error('No active logger found')
  }

  if (logger instanceof Logger) {
    return logger
  }

  throw new Error('Active logger is not an instance of Logger')
}

/**
 * Runs a function with a new OpenTelemetry context containing the provided logger.
 *
 * @param logger The logger to make active
 * @param fn The function to run with the logger in context
 * @returns The return value of the function
 */
export function withActiveLogger<T>(logger: Logger, fn: () => T): T {
  return context.with(context.active().setValue(LOG_CONTEXT, logger), fn)
}
