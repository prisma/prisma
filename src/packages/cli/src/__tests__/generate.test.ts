import 'ts-node/register'
import { consoleContext, Context } from './__helpers__/context'

const ctx = Context.new().add(consoleContext()).assemble()

it('should work with a custom output dir', async () => {
  ctx.fixture('example-project')
  const data = await ctx.cli('generate')

  if (typeof data.signal === 'number' && data.signal !== 0) {
    throw new Error(data.stderr + data.stdout)
  }

  const { main } = await import(ctx.fs.path('main.ts'))
  expect(cleanSnapshot(data.stdout)).toMatchSnapshot()
  await expect(main()).resolves.toMatchSnapshot()
}, 10000)

it('should error with exit code 1 with incorrect schema', async () => {
  ctx.fixture('broken-example-project')
  await expect(ctx.cli('generate').catch((e) => e.exitCode)).resolves.toEqual(1)
})

function cleanSnapshot(str: string): string {
  return str
    .replace(/\d+ms/g, 'XXms')
    .replace(/\d+s/g, 'XXs')
    .replace(/\(version:.+\)/g, '(version: 0.0.0)')
}
