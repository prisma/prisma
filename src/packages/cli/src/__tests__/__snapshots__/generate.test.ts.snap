// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should work with a custom output dir 1`] = `

Prisma schema loaded from prisma/schema.prisma

âœ” Generated Prisma Client (2.11.0-dev.20) to ./generated/client in XXms
You can now start using Prisma Client in your code. Reference: https://pris.ly/d/client
\`\`\`
import { PrismaClient } from './generated/client'
const prisma = new PrismaClient()
\`\`\`

warn Prisma 2.12.0 has breaking changes.
You can update your code with npx @prisma/codemods update-2.12 ./
Read more at https://pris.ly/2.12
`;

exports[`should work with a custom output dir 2`] = `
Array [
  Object {
    email: bob@bob.bob,
    id: 1,
    name: Bobby Brown Sqlite,
  },
]
`;
